Cannot parse HGVS `NP_000095.2:p.Gly168_His169=`

When trying to parse `NP_000095.2:p.Gly168_His169=` I get:

```
Caused by: de.charite.compbio.jannovar.hgvs.parser.HGVSParsingException: org.antlr.v4.runtime.NoViableAltException
	at de.charite.compbio.jannovar.hgvs.parser.HGVSErrorStrategy.reportNoViableAlternative(HGVSErrorStrategy.java:11)
	at org.antlr.v4.runtime.DefaultErrorStrategy.reportError(DefaultErrorStrategy.java:136)
	at de.charite.compbio.jannovar.hgvs.parser.Antlr4HGVSParser.aa_change_inner(Antlr4HGVSParser.java:888)
	at de.charite.compbio.jannovar.hgvs.parser.Antlr4HGVSParser.aa_change(Antlr4HGVSParser.java:743)
	at de.charite.compbio.jannovar.hgvs.parser.Antlr4HGVSParser.aa_single_allele_single_change_var(Antlr4HGVSParser.java:372)
	at de.charite.compbio.jannovar.hgvs.parser.Antlr4HGVSParser.aa_single_allele_var(Antlr4HGVSParser.java:316)
	at de.charite.compbio.jannovar.hgvs.parser.Antlr4HGVSParser.hgvs_variant(Antlr4HGVSParser.java:201)
	at de.charite.compbio.jannovar.hgvs.parser.HGVSParser.parseHGVSString(HGVSParser.java:42)
	... 75 common frames omitted
Caused by: org.antlr.v4.runtime.NoViableAltException: null
	at org.antlr.v4.runtime.atn.ParserATNSimulator.noViableAlt(ParserATNSimulator.java:2026)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.execATN(ParserATNSimulator.java:467)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.adaptivePredict(ParserATNSimulator.java:393)
	at de.charite.compbio.jannovar.hgvs.parser.Antlr4HGVSParser.aa_change_inner(Antlr4HGVSParser.java:820)
	... 81 common frames omitted
```

This is with Janovar 0.34.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,10 @@
   The deprecated classes are due to be removed in v0.36.
 * Adding modules for importing VCF files into H2 database files, listing the meta data conents, and annotating VCF files.
 
+### jannovar-hgvs
+
+* Fixing parsing of unchanged (`=`) for nucleic acide sequences (#493).
+
 ## v0.34
 
 ### jannovar-cli

--- a/jannovar-hgvs/src/main/antlr4/de/charite/compbio/jannovar/hgvs/parser/Antlr4HGVSParser.g4
+++ b/jannovar-hgvs/src/main/antlr4/de/charite/compbio/jannovar/hgvs/parser/Antlr4HGVSParser.g4
@@ -115,6 +115,7 @@ aa_change_inner
 	| aa_change_substitution
 	| aa_change_ssr
 	| aa_change_insertion
+	| aa_change_unchanged
 	| aa_change_misc
 ;
 
@@ -129,6 +130,16 @@ aa_change_deletion
 		AA_NUMBER
 		| aa_string
 	)?
+;
+
+
+/** amino acid unchanged */
+aa_change_unchanged
+:
+	(
+		aa_point_location
+		| aa_range
+	) AA_EQUAL
 ;
 
 /** amino acid duplication */
@@ -373,6 +384,7 @@ nt_change_inner
 	| nt_change_inversion
 	| nt_change_substitution
 	| nt_change_ssr
+	| nt_change_unchanged
 	| nt_change_misc
 ;
 
@@ -400,6 +412,18 @@ nt_change_duplication
 		nt_number
 		| nt_string
 	)?
+;
+
+/** unchanged nucleotides */
+nt_change_unchanged
+:
+	(
+		(
+		    nt_point_location
+		    nt_string
+		)
+		| nt_range
+	) NT_EQUAL
 ;
 
 /** nucleotide replacement/indel/delins */

--- a/jannovar-hgvs/src/main/java/de/charite/compbio/jannovar/hgvs/nts/change/NucleotideUnchanged.java
+++ b/jannovar-hgvs/src/main/java/de/charite/compbio/jannovar/hgvs/nts/change/NucleotideUnchanged.java
@@ -1,0 +1,127 @@
+package de.charite.compbio.jannovar.hgvs.nts.change;
+
+import com.google.common.base.Joiner;
+import de.charite.compbio.jannovar.hgvs.nts.NucleotideRange;
+import de.charite.compbio.jannovar.hgvs.nts.NucleotideSeqDescription;
+
+/**
+ * Unchanged in a nucleotide sequence.
+ *
+ * @author <a href="mailto:manuel.holtgrewe@bihealth.de">Manuel Holtgrewe</a>
+ */
+public class NucleotideUnchanged extends NucleotideChange {
+
+	/**
+	 * unchanged range of nucleotides
+	 */
+	private final NucleotideRange range;
+	/**
+	 * description of the deleted nucleotide sequence
+	 */
+	private final NucleotideSeqDescription seq;
+
+	public static NucleotideUnchanged buildWithOffset(boolean onlyPredicted, int firstPos, int firstPosOffset,
+                                                      int lastPos, int lastPosOffset, NucleotideSeqDescription seq) {
+		return new NucleotideUnchanged(onlyPredicted, NucleotideRange.build(firstPos, firstPosOffset, lastPos,
+			lastPosOffset), seq);
+	}
+
+	public static NucleotideUnchanged buildWithOffsetWithSequence(boolean onlyPredicted, int firstPos,
+                                                                  int firstPosOffset, int lastPos, int lastPosOffset, String nts) {
+		return new NucleotideUnchanged(onlyPredicted, NucleotideRange.build(firstPos, firstPosOffset, lastPos,
+			lastPosOffset), new NucleotideSeqDescription(nts));
+	}
+
+	public static NucleotideUnchanged buildWithOffsetWithLength(boolean onlyPredicted, int firstPos, int firstPosOffset,
+                                                                int lastPos, int lastPosOffset, int seqLen) {
+		return new NucleotideUnchanged(onlyPredicted, NucleotideRange.build(firstPos, firstPosOffset, lastPos,
+			lastPosOffset), new NucleotideSeqDescription(seqLen));
+	}
+
+	public static NucleotideUnchanged buildWithOffsetWithoutSeqDescription(boolean onlyPredicted, int firstPos,
+                                                                           int firstPosOffset, int lastPos, int lastPosOffset) {
+		return new NucleotideUnchanged(onlyPredicted, NucleotideRange.build(firstPos, firstPosOffset, lastPos,
+			lastPosOffset), new NucleotideSeqDescription());
+	}
+
+	public static NucleotideUnchanged build(boolean onlyPredicted, int firstPos, int lastPos,
+                                            NucleotideSeqDescription seq) {
+		return new NucleotideUnchanged(onlyPredicted, NucleotideRange.buildWithoutOffset(firstPos, lastPos), seq);
+	}
+
+	public static NucleotideUnchanged buildWithSequence(boolean onlyPredicted, int firstPos, int lastPos, String nts) {
+		return new NucleotideUnchanged(onlyPredicted, NucleotideRange.buildWithoutOffset(firstPos, lastPos),
+			new NucleotideSeqDescription(nts));
+	}
+
+	public static NucleotideUnchanged buildWithLength(boolean onlyPredicted, int firstPos, int lastPos, int seqLen) {
+		return new NucleotideUnchanged(onlyPredicted, NucleotideRange.buildWithoutOffset(firstPos, lastPos),
+			new NucleotideSeqDescription(seqLen));
+	}
+
+	public static NucleotideUnchanged buildWithoutSeqDescription(boolean onlyPredicted, int firstPos, int lastPos) {
+		return new NucleotideUnchanged(onlyPredicted, NucleotideRange.buildWithoutOffset(firstPos, lastPos),
+			new NucleotideSeqDescription());
+	}
+
+	public NucleotideUnchanged(boolean onlyPredicted, NucleotideRange range, NucleotideSeqDescription seq) {
+		super(onlyPredicted);
+		this.range = range;
+		this.seq = seq;
+	}
+
+	@Override
+	public NucleotideUnchanged withOnlyPredicted(boolean flag) {
+		return new NucleotideUnchanged(flag, range, seq);
+	}
+
+	public NucleotideRange getRange() {
+		return range;
+	}
+
+	public NucleotideSeqDescription getSeq() {
+		return seq;
+	}
+
+	@Override
+	public String toHGVSString() {
+		return wrapIfOnlyPredicted(Joiner.on("").join(range.toHGVSString(), seq.toHGVSString(), "="));
+	}
+
+	@Override
+	public String toString() {
+		return "NucleotideUnchanged [range=" + range + ", seq=" + seq + "]";
+	}
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = super.hashCode();
+		result = prime * result + ((range == null) ? 0 : range.hashCode());
+		result = prime * result + ((seq == null) ? 0 : seq.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (!super.equals(obj))
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		NucleotideUnchanged other = (NucleotideUnchanged) obj;
+		if (range == null) {
+			if (other.range != null)
+				return false;
+		} else if (!range.equals(other.range))
+			return false;
+		if (seq == null) {
+			if (other.seq != null)
+				return false;
+		} else if (!seq.equals(other.seq))
+			return false;
+		return true;
+	}
+
+}

--- a/jannovar-hgvs/src/main/java/de/charite/compbio/jannovar/hgvs/parser/Antlr4HGVSParserListenerImpl.java
+++ b/jannovar-hgvs/src/main/java/de/charite/compbio/jannovar/hgvs/parser/Antlr4HGVSParserListenerImpl.java
@@ -318,6 +318,31 @@ class Antlr4HGVSParserListenerImpl extends Antlr4HGVSParserBaseListener {
 		setValue(ctx, change);
 	}
 
+
+	/**
+	 * Leaving of nt_change_unchanged rule
+	 * <p>
+	 * Construct {@link NucleotideUnchanged} from children's values and labels and label ctx with
+	 * this.
+	 */
+	@Override
+	public void exitNt_change_unchanged(Nt_change_unchangedContext ctx) {
+		LOGGER.debug("Leaving nt_change_unchanged");
+		final NucleotideRange range;
+		if (ctx.nt_range() != null)
+			range = (NucleotideRange) getValue(ctx.nt_range());
+		else
+			range = new NucleotideRange((NucleotidePointLocation) getValue(ctx.nt_point_location()),
+				(NucleotidePointLocation) getValue(ctx.nt_point_location()));
+		final NucleotideUnchanged change;
+		if (ctx.nt_string() != null)
+			change = new NucleotideUnchanged(false, range,
+				new NucleotideSeqDescription(ctx.nt_string().getText()));
+		else
+			change = new NucleotideUnchanged(false, range, new NucleotideSeqDescription());
+		setValue(ctx, change);
+	}
+
 	/**
 	 * Leaving of nt_change_duplication rule
 	 * <p>

--- a/jannovar-hgvs/src/main/java/de/charite/compbio/jannovar/hgvs/protein/change/ProteinUnchanged.java
+++ b/jannovar-hgvs/src/main/java/de/charite/compbio/jannovar/hgvs/protein/change/ProteinUnchanged.java
@@ -1,0 +1,106 @@
+package de.charite.compbio.jannovar.hgvs.protein.change;
+
+import de.charite.compbio.jannovar.hgvs.AminoAcidCode;
+import de.charite.compbio.jannovar.hgvs.protein.ProteinRange;
+import de.charite.compbio.jannovar.hgvs.protein.ProteinSeqDescription;
+
+// TODO(holtgrewe): Remove seqDesc?
+
+/**
+ * Unchanged mark inside protein.
+ *
+ * @author <a href="mailto:manuel.holtgrewe@bihealth.de">Manuel Holtgrewe</a>
+ */
+public class ProteinUnchanged extends ProteinChange {
+
+	/**
+	 * range of one or more amino acids that are deleted
+	 */
+	private final ProteinRange range;
+	/**
+	 * specification of the deleted characters, can be null
+	 */
+	private final ProteinSeqDescription seqDesc;
+
+	/**
+	 * Construct ProteinDeletion without length and sequence information
+	 */
+	public static ProteinUnchanged buildWithoutSeqDescription(boolean onlyPredicted, String firstAA, int firstPos,
+                                                              String lastAA, int lastPos) {
+		return new ProteinUnchanged(onlyPredicted, ProteinRange.build(firstAA, firstPos, lastAA, lastPos));
+	}
+
+	/**
+	 * Construct ProteinDeletion with length information
+	 */
+	public static ProteinUnchanged buildWithLength(boolean onlyPredicted, String firstAA, int firstPos,
+                                                   String lastAA, int lastPos, int length) {
+		return new ProteinUnchanged(onlyPredicted, ProteinRange.build(firstAA, firstPos, lastAA, lastPos), length);
+	}
+
+	/**
+	 * Construct ProteinDeletion with sequence
+	 */
+	public static ProteinUnchanged buildWithSequence(boolean onlyPredicted, String firstAA, int firstPos, String lastAA,
+                                                     int lastPos, String seq) {
+		return new ProteinUnchanged(onlyPredicted, ProteinRange.build(firstAA, firstPos, lastAA, lastPos), seq);
+	}
+
+	/**
+	 * Construct ProteinDeletion without length and sequence information
+	 */
+	public ProteinUnchanged(boolean onlyPredicted, ProteinRange range) {
+		super(onlyPredicted);
+		this.range = range;
+		this.seqDesc = new ProteinSeqDescription();
+	}
+
+	/**
+	 * Construct ProteinDeletion with length information
+	 */
+	public ProteinUnchanged(boolean onlyPredicted, ProteinRange range, int length) {
+		super(onlyPredicted);
+		this.range = range;
+		this.seqDesc = new ProteinSeqDescription(length);
+	}
+
+	/**
+	 * Construct ProteinDeletion with sequence information
+	 */
+	public ProteinUnchanged(boolean onlyPredicted, ProteinRange range, String seq) {
+		super(onlyPredicted);
+		this.range = range;
+		this.seqDesc = new ProteinSeqDescription(seq);
+	}
+
+	private ProteinUnchanged(boolean onlyPredicted, ProteinRange range, ProteinSeqDescription seqDesc) {
+		super(onlyPredicted);
+		this.range = range;
+		this.seqDesc = seqDesc;
+	}
+
+	/**
+	 * @return deleted range in the protein
+	 */
+	public ProteinRange getRange() {
+		return range;
+	}
+
+	/**
+	 * @return description of the deleted sequence
+	 */
+	public ProteinSeqDescription getSeqDesc() {
+		return seqDesc;
+	}
+
+	@Override
+	public String toHGVSString(AminoAcidCode code) {
+		return wrapIfOnlyPredicted(range.toHGVSString(code) + "=");
+	}
+
+	@Override
+	public ProteinChange withOnlyPredicted(boolean onlyPredicted) {
+		return new ProteinUnchanged(onlyPredicted, this.range, this.seqDesc);
+	}
+
+}

--- a/jannovar-hgvs/src/test/java/de/charite/compbio/jannovar/hgvs/parser/HGVSParserDriverNucleotideMiscChangeTest.java
+++ b/jannovar-hgvs/src/test/java/de/charite/compbio/jannovar/hgvs/parser/HGVSParserDriverNucleotideMiscChangeTest.java
@@ -23,7 +23,7 @@ public class HGVSParserDriverNucleotideMiscChangeTest {
 	@Test
 	public void testParseString() {
 		String hgvsStrings[] = new String[]{"XXX:r.(?)", "XXX:r.?", "XXX:r.spl?", "XXX:r.(spl?)", "XXX:r.=",
-			"XXX:r.(=)", "XXX:r.0", "XXX:r.(0)"};
+			"XXX:r.(=)", "XXX:r.0", "XXX:r.(0)", "XXX:c.97A=", "XXX:c.79_97="};
 
 		for (String hgvsString : hgvsStrings) {
 			HGVSVariant variant = driver.parseHGVSString(hgvsString);

--- a/jannovar-hgvs/src/test/java/de/charite/compbio/jannovar/hgvs/parser/nts/NucleotideParsingTest.java
+++ b/jannovar-hgvs/src/test/java/de/charite/compbio/jannovar/hgvs/parser/nts/NucleotideParsingTest.java
@@ -21,4 +21,13 @@ public class NucleotideParsingTest extends HGVSParsingTestBase {
 				parseString(PREFIX + t + s);
 	}
 
+	@Test
+	public void testNucleotideSingleVarUnchanged() {
+		String[] types = {"c.", "m.", "n.", "g.", "r."};
+		String[] changes = {"76A=", "67_76="};
+		for (String t : types)
+			for (String s : changes)
+				parseString(PREFIX + t + s);
+	}
+
 }

--- a/jannovar-hgvs/src/test/java/de/charite/compbio/jannovar/hgvs/parser/protein/ProteinParsingTest.java
+++ b/jannovar-hgvs/src/test/java/de/charite/compbio/jannovar/hgvs/parser/protein/ProteinParsingTest.java
@@ -70,6 +70,13 @@ public class ProteinParsingTest extends HGVSParsingTestBase {
 	}
 
 	@Test
+	public void testProteinUnchanged() {
+		String[] arr = {"p.G33=", "p.Arg97_Cys100="};
+		for (String s : arr)
+			parseString(PREFIX + s);
+	}
+
+	@Test
 	public void testProteinSingleLongFrameShift() {
 		String[] arr = {"p.Arg97Profs*23", "p.A97Pfs*23"};
 		for (String s : arr)

